File: src/client/Systems/Units/RemoveMissingModels.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local Packages = ReplicatedStorage.Packages
local Matter = require(Packages.Matter)

local Components = require(ReplicatedStorage.Source.Components)

type ModelComponent = Components.ModelComponent
type World = Matter.World

--[=[
    @within Unit
    @tag System
    @client
    
    Handles the removal of the model component from units that are no longer in
    the game, along with removing the physical model from the game for units
    that are no longer in the world.
]=]
local function RemoveMissingModels(world: World)
    for id: number, model: ModelComponent in world:query(Components.Model) do
        for _ in Matter.useEvent(model.model, "AncestryChanged") do
            if model.model:IsDescendantOf(game) == false then
                world:remove(id, Components.Model)
                break
            end
        end
        if not model.model.PrimaryPart then
            world:remove(id, Components.Model)
        end
    end

    for _id: number, modelRecord in world:queryChanged(Components.Model) do
        if modelRecord.new == nil then
            if modelRecord.old and modelRecord.old.model then
                modelRecord.old.model:Destroy()
            end
        end
    end
end

return RemoveMissingModels

File: src/client/systems/units/renderable/remove-missing-models.ts
import { useEvent, World } from "@rbxts/matter";
import * as Components from "shared/components";

/**
 * Handles the removal of the model component from units that are no longer in
 * the game, along with removing the physical model from the game for units
 * that are no longer in the world.
 *
 * @param world from Matter
 */
function removeMissingModels(world: World): void {
	for (const [id, model] of world.query(Components.Renderable)) {
		for (const [_] of useEvent(model.model, model.model.AncestryChanged)) {
			if (!model.model.IsDescendantOf(game)) {
				world.remove(id, Components.Renderable);
				break;
			}
		}
		if (!model.model.PrimaryPart) {
			world.remove(id, Components.Renderable);
		}
	}

	for (const [, modelRecord] of world.queryChanged(Components.Renderable)) {
		if (modelRecord.new !== undefined) {
			continue;
		}
		if (modelRecord.old) {
			modelRecord.old.model.Destroy();
		}
	}
}

export = removeMissingModels;
